### PR TITLE
Multiple plans functions

### PR DESCRIPTION
**Pull request by elizabeth-tang**
Summary
This pull request will support the backend for the multiple plans feature

- Add new data type: Plan that consists of a name for the plan and an array of semesters
- Update Vuex to keep track of plans and the current plan
- Add Firestore functions and Vuex mutations to support adding, deleting and modifying plans

### Test Plan 
- Unit tests using Jest
- Eventually connecting to the frontend after modals are created

### CODE REVIEW

File: src/user-data.d.ts
Comment (zachary0kent) [inline src/user-data.d.ts:1]: Would it make sense for the changes here to be in the migration script PR instead?

File: src/global-firestore-data/user-semesters.ts [hunk 1]
```diff
@@ -23,6 +23,14 @@ export const editSemesters = (
...
 };
 
+export const editPlans = (updater: (oldPlans: readonly Plan[]) => readonly Plan[]): void => {
+  const plans = updater(store.state.plans);
+  store.commit('setPlans', plans);
+  updateDoc(doc(semestersCollection, store.state.currentFirebaseUser.email), {
+    plans,
+  });
+};
+
```
Comment (zachary0kent) [inline src/global-firestore-data/user-semesters.ts:33]:
```suggestion
export const editPlans = async (updater: (oldPlans: readonly Plan[]) => readonly Plan[]): void => {
  const plans = updater(store.state.plans);
  store.commit('setPlans', plans);
  await updateDoc(doc(semestersCollection, store.state.currentFirebaseUser.email), {
    plans,
  });
};
```
`updateDoc` returns a promise that should resolve before `editPlans` returns

File: src/global-firestore-data/user-semesters.ts [hunk 2]
```diff
@@ -75,6 +98,11 @@ export const addSemester = (
   editSemesters(oldSemesters => [...oldSemesters, createSemester(year, season, courses)]);
 };
 
+export const addPlan = (name: string, semesters: FirestoreSemester[], gtag?: GTag): void => {
+  GTagEvent(gtag, 'add-plan');
+  editPlans(oldPlans => [...oldPlans, createPlan(name, semesters)]);
+};
```
Comment (zachary0kent) [inline src/global-firestore-data/user-semesters.ts:104]:
```suggestion
export const addPlan = async (name: string, semesters: FirestoreSemester[], gtag?: GTag): void => {
  GTagEvent(gtag, 'add-plan');
  await editPlans(oldPlans => [...oldPlans, createPlan(name, semesters)]);
};
```

File: src/global-firestore-data/user-semesters.ts [hunk 3]
```diff
@@ -88,6 +116,14 @@ export const deleteSemester = (
   }
 };
 
+export const deletePlan = (name: string, gtag?: GTag): void => {
+  GTagEvent(gtag, 'delete-plan');
+  const plan = store.state.plans.find(p => p.name === name);
+  if (plan) {
+    editPlans(oldPlans => oldPlans.filter(p => !(p.name === name)));
+  }
+};
```
Comment (zachary0kent) [inline src/global-firestore-data/user-semesters.ts:125]:
```suggestion
export const deletePlan = async (name: string, gtag?: GTag): void => {
  GTagEvent(gtag, 'delete-plan');
  if (store.state.plans.some(p => p.name === name)) {
    await editPlans(oldPlans => oldPlans.filter(p => p.name !== name));
  }
};
```

File: src/store.ts [hunk 4]
```diff
@@ -157,6 +161,12 @@ const store: TypedVuexStore = new TypedVuexStore({
     setIsTeleportModalOpen(state: VuexStoreState, newTeleportModalValue: boolean) {
       state.isTeleportModalOpen = newTeleportModalValue;
     },
+    setPlans(state: VuexStoreState, newPlans: Plan[]) {
```
Comment (zachary0kent) [inline src/store.ts:164]:
```suggestion
    setPlans(state: VuexStoreState, newPlans: readonly Plan[]) {
```
This tells the type system that `setPlans` won't modify `newPlans`

**Review by zachary0kent — CHANGES_REQUESTED** [review]
Thanks for adding analytics events for adding/removing a plan! Everything looks good overall, just left a couple comments about async/await and a couple nits.

File: src/components/Requirements/MultiplePlansDropdown.vue [hunk 5]
```diff
@@ -37,14 +37,15 @@
...
+import store from '@/store';
 
 export default defineComponent({
   props: {
     plan: { type: String, default: 'No additional plans yet' },
   },
   data() {
     return {
-      plans: ['PLAN 1', 'PLAN 2', 'PLAN 3', 'PLAN 4'],
+      plans: store.state.plans.map(plan => plan.name),
```
Comment (zachary0kent) [inline src/components/Requirements/MultiplePlansDropdown.vue:48]: Now that the user's plans are maintained in Firestore, we should probably make `plans` a computed property instead. Right now, deleting or adding a plan would not be reflected by the dropdown.

**Review by zachary0kent — CHANGES_REQUESTED** [review]

**Review by PabloRaigoza — COMMENTED** [review]
Good work Elizabeth! I liked it how you incorporated analytics checking for when users added and removed plans.
Maybe keeping track of how often a user switches to a different plan could be helpful as it could let us see how useful this feature is.

File: src/components/Requirements/MultiplePlansDropdown.vue (hunk 5 above)
Comment (noschiff) [reply src/components/Requirements/MultiplePlansDropdown.vue:48]: Zak's a Vue wizard 🧙🏻‍♂️

**Review by zachary0kent — APPROVED** [review]
Looks good!

**Pull request merged**